### DISCUSSION TITLE
Form submissions doesn't reset orders state

### DISCUSSION THREAD

**Discussion by Gabrielle-Tobermann** [nss-evening-cohort-14/evening-client-side#315] [Capstones!] [answered]
Gabby 

Other tickets:  #287 #306

Issue: When I submit my form for a new order, I can a few errors. I've got a promise.all that creates an order and creates orderItems is a different table. When the form is submitted, my data is successfully being pushed to firebase. However, I need to hard refresh for the form to disappear and for the new orders to show up. After I hard refresh, they appear. On submission, I get the following errors: 

```
utils.js:248 Uncaught TypeError: Cannot read property 'length' of undefined
    at isArrayOrNodeList (utils.js:248)
    at getTarget (utils.js:254)
    at TooltipPopoverWrapper.updateTarget
```
```
index.js:1 Warning: Each child in a list should have a unique "key" prop.
```
I definitely have a key prop. The value is the firebaseKey of the order. 
```
Check the render method of OrderView.Uncaught (in promise) TypeError: Cannot read property 'length' of undefined
    at isArrayOrNodeList
```

What I have tried: I thought maybe it was because of my promise.all, but I tried just creating orders (just the createOrder promise) but the errors still remain. I made sure of setOrders (state) to the response of the promise. I made sure I was calling getOrders() in the .then() of my promise. I followed the exact same process as for my items, which I didn't have an issue with. I declared my orders hook in the Orders view, because I don't need it anywhere else, unlike my items. I believe I have my hooks in the right place. When I refresh the page, my new orders are there. I thought maybe my component was rerendering before my promise completes, because when I put  debugger in my handleSubmit function, it does not throw an error. But, since I tested it with just a simple createOrder promise and the same thing happens, I assume this is not the case. This is also just happening with create. My delete functionality is working as it should. It seems like my setOrders hook is not updating after the order is created, even though I am calling setOrders and passing the response from the promise to it.

Google links: 
https://stackoverflow.com/questions/42132290/wait-for-react-promise-to-resolve-before-render
https://stackoverflow.com/questions/53078297/form-is-not-reset-after-submit-react-js/53078478

Links to code: 
https://github.com/Gabrielle-Tobermann/Overture-1/blob/orders-read-create-delete/src/components/OrderForm.js
https://github.com/Gabrielle-Tobermann/Overture-1/blob/orders-read-create-delete/src/views/OrderView.js
https://github.com/Gabrielle-Tobermann/Overture-1/blob/orders-read-create-delete/src/helpers/data/ordersData.js
https://github.com/Gabrielle-Tobermann/Overture-1/blob/orders-read-create-delete/src/components/OrderCard.js

**Comment by drteresavasquez** [comment]
I don't see the links to the code in your ticket. Also, what do you have .length on? Because it's not defined.

**Reply by Gabrielle-Tobermann** [reply]
My bad, I've added my links.

**Reply by Gabrielle-Tobermann** [reply]
I actually don't have .length on anything, I think the .length of undefined is coming from the popover. It says the error is at ```node_modules/reactstrap/es/utils.js:248``` The weird thing is, it's being inconsistent. Sometimes I can add new orders and it will update the orders without throwing an error, but the form is still there with the values I entered. The form never disappears. And despite the errors, when I refresh the page, my new orders are always adding.

**Comment by Gabrielle-Tobermann** [comment]
Update: I moved my getOrders call and my orders hook to index.js and now it seems to be working properly most of the time (the DOM updates with new orders and the form disappears). BUT, sometimes, it will throw the same errors again. It will work like 5 times in a row and them throw same errors again.

**Reply by drteresavasquez** [reply]
Why do you need to setOrders if you will make a call to the API for the actual data? We may need to look in class tomorrow so I can better understand the issue.

**Reply by Gabrielle-Tobermann** [reply]
I need setOrders because I have an orders view which displays orders by customer name, and details about the order in a popover, name, email. date of the order, and items purchased (the items are in a seperate table in firebase but are displayed on the popover). And since I'm not quite done with MVP, I am not doing the stripe stuff yet. If I can't manage to get stripe working, I still want to show something with the orders view.

**Comment by TrinityChristiana** [comment]
@Gabrielle-Tobermann Will add the solution!

**Comment by Gabrielle-Tobermann** [comment, marked as answer]
Solution: In my promise, I had my .then(getOrders()) in the wrong place. I needed to attach it to the patch. It was getting the orders before it had a chance to patch the firebaseKey.
